File: MiniDOMOffice/Magazzino/Inventario/AttributiOggetto/Collection/AttributiOggettoCollection.cs
using System;
using System.Collections;
using System.Collections.Generic;
using DMD;
using DMD.XML;
using DMD.Databases;
using DMD.Databases.Collections;
using minidom;
using minidom.repositories;
using static minidom.Sistema;
using static minidom.Anagrafica;
using static minidom.Store;


namespace minidom
{
    public partial class Store
    {

        /// <summary>
        /// Collezione degli attributi specificato per una istanza di un Oggetto
        /// </summary>
        /// <remarks></remarks>
        [Serializable]
        public class AttributiOggettoCollection
            : CKeyCollection<AttributoOggetto>
        {
            [NonSerialized]
            private OggettoInventariato m_Oggetto;

            /// <summary>
            /// Costruttore
            /// </summary>
            public AttributiOggettoCollection()
            {
                m_Oggetto = null;
            }


            /// <summary>
            /// Costruttore
            /// </summary>
            /// <param name="Oggetto"></param>
            public AttributiOggettoCollection(OggettoInventariato Oggetto) 
                : this()
            {
                Load(Oggetto);
            }

            /// <summary>
            /// Restitusice un riferimento all'oggetto inventariato
            /// </summary>
            public OggettoInventariato Oggetto
            {
                get
                {
                    return m_Oggetto;
                }
            }

            /// <summary>
            /// Imposta il riferimento all'oggetto inventariato
            /// </summary>
            /// <param name="value"></param>
            protected internal virtual void SetOggetto(OggettoInventariato value)
            {
                m_Oggetto = value;
                if (value is object)
                {
                    foreach (AttributoOggetto item in this)
                        item.SetOggetto(value);
                }
            }

            /// <summary>
            /// OnInsert
            /// </summary>
            /// <param name="index"></param>
            /// <param name="value"></param>
            protected override void OnInsert(int index, AttributoOggetto value)
            {
                if (m_Oggetto is object)
                    value.SetOggetto(m_Oggetto);
                base.OnInsert(index, value);
            }

            /// <summary>
            /// OnSet
            /// </summary>
            /// <param name="index"></param>
            /// <param name="oldValue"></param>
            /// <param name="newValue"></param>
            protected override void OnSet(int index, AttributoOggetto oldValue, AttributoOggetto newValue)
            {
                if (m_Oggetto is object)
                    newValue.SetOggetto(m_Oggetto);
                base.OnSet(index, oldValue, newValue);
            }

            /// <summary>
            /// Carica la collezione
            /// </summary>
            /// <param name="Oggetto"></param>
            protected internal virtual void Load(OggettoInventariato Oggetto)
            {
                if (Oggetto is null)
                    throw new ArgumentNullException("Oggetto");
                
                Clear();
                
                m_Oggetto = Oggetto;

                if (DBUtils.GetID(Oggetto, 0) == 0)
                    return;

                using (var cursor = new AttributoOggettoCursor())
                {
                    cursor.IgnoreRights = true;
                    cursor.IDOggetto.Value = DBUtils.GetID(Oggetto, 0);
                    while (cursor.Read())
                    {
                        Add(cursor.Item.NomeAttributo, cursor.Item);

                    }
                }

                Sort();
            }
        }
    }
}

File: MiniDOMOffice/Magazzino/Inventario/AttributiOggetto/AttributiOggettoCollection.vb
Imports minidom.Databases
Imports minidom.Sistema
Imports minidom.Anagrafica



Partial Public Class Office
     
    ''' <summary>
    ''' Collezione degli attributi specificato per una istanza di un Oggetto
    ''' </summary>
    ''' <remarks></remarks>
    <Serializable> _
    Public Class AttributiOggettoCollection
        Inherits CKeyCollection(Of AttributoOggetto)

        <NonSerialized> Private m_Oggetto As OggettoInventariato

        Public Sub New()
            Me.m_Oggetto = Nothing
        End Sub

        Public Sub New(ByVal Oggetto As OggettoInventariato)
            Me.New()
            Me.Load(Oggetto)
        End Sub

        Public ReadOnly Property Oggetto As OggettoInventariato
            Get
                Return Me.m_Oggetto
            End Get
        End Property

        Protected Friend Overridable Sub SetOggetto(ByVal value As OggettoInventariato)
            Me.m_Oggetto = value
            If (value IsNot Nothing) Then
                For Each item As AttributoOggetto In Me
                    item.SetOggetto(value)
                Next
            End If
        End Sub

        Protected Overrides Sub OnInsert(index As Integer, value As Object)
            If (Me.m_Oggetto IsNot Nothing) Then DirectCast(value, AttributoOggetto).SetOggetto(Me.m_Oggetto)
            MyBase.OnInsert(index, value)
        End Sub

        Protected Overrides Sub OnSet(index As Integer, oldValue As Object, newValue As Object)
            If (Me.m_Oggetto IsNot Nothing) Then DirectCast(newValue, AttributoOggetto).SetOggetto(Me.m_Oggetto)
            MyBase.OnSet(index, oldValue, newValue)
        End Sub

        Public Sub Load(ByVal Oggetto As OggettoInventariato)
            If (Oggetto Is Nothing) Then Throw New ArgumentNullException("Oggetto")
            Me.Clear()
            Me.m_Oggetto = Oggetto
            If (GetID(Oggetto) = 0) Then Exit Sub
            Dim cursor As New AttributoOggettoCursor
            cursor.IgnoreRights = True
            cursor.IDOggetto.Value = GetID(Oggetto)
            While Not cursor.EOF
                Me.Add(cursor.Item.NomeAttributo, cursor.Item)
                cursor.MoveNext()
            End While
            cursor.Dispose()
            Me.Sort()
        End Sub

    End Class

End Class


